bigml/anomaly.py

```diff
@@ -15,131 +15,202 @@
-from bigml.util import cast
+from bigml.api import get_status, get_api_connection, get_anomaly_id
 from bigml.basemodel import get_resource_dict
 from bigml.modelfields import ModelFields
-from bigml.anomalytree import AnomalyTree
-
+from bigml.util import cast
 
-LOGGER = logging.getLogger('BigML')
 
 DEPTH_FACTOR = 0.5772156649
 
 
-def norm_factor(sample_size, mean_depth):
-    """Computing the normalization factor for simple anomaly detectors
+def build_tree(children, node=None):
+    """Builds a compressed version of the tree structure as an list of
+    lists. Starting from the root node, that is represented by a list:
+        [len(children), children1, children2, etc.]
+    And each child is represented by a list whose elements are:
+        [weight, len(predicates), operator_code, field, value, term, missing,
+         ..., len(children), children_nodes_list*]
+    """
+    outer = node if node else list()
+    outer.append(len(children))
+    for child in children:
+        inner = list()
+        inner.append(child.get('weight'))
+        predicates = child.get('predicates')
+        if predicates and not (predicates is True or predicates == [True]):
+            predicates = [x for x in predicates if x is not True]
+            inner.append(len(predicates))
+            for pred in predicates:
+                operation = pred.get('op')
+                value = pred.get('value')
+                missing = False
+                if operation.endswith("*"):
+                    operation = operation[0: -1]
+                    missing = True
+                elif operation == 'in' and None in value:
+                    missing = True
+
+                inner.append(OPERATOR_CODE.get(operation))
+                inner.append(pred.get('field'))
+                inner.append(value)
+                inner.append(pred.get('term'))
+                inner.append(missing)
+        else:
+            inner.append(0)
+        next_gen = child.get('children')
+        if next_gen:
+            build_tree(next_gen, node=inner)
+        else:
+            inner.append(0)
+        outer.append(inner)
+
+    return outer
+
 
+def calculate_depth(node, input_data, fields, depth=0):
+    """Computes the depth in the tree for the input data
     """
-    if mean_depth is not None:
-        default_depth = mean_depth if sample_size == 1 else \
-            (2 * (DEPTH_FACTOR + math.log(sample_size - 1) - \
-            (float(sample_size - 1) / sample_size)))
-        return min(mean_depth, default_depth)
 
+    weight = node[0] if node[0] else 1
+    num_predicates = node[1]
+    num_children = node[2 + (5 * num_predicates)]
 
-class Anomaly(ModelFields):
-    """ A lightweight wrapper around an anomaly detector.
+    if not apply_predicates(node, input_data, fields):
+        return depth
+
+    depth += weight
+    if num_children > 0:
+        start = 3 + (5 * num_predicates)
+        end = 3 + num_children + (5 * num_predicates)
```

Comment:
the value `5 * num_predicates` appears in at least three places: maybe we could define a variable? (if it can be given a reasonable name, that magic constant 3 could also be named)